File: lib/collection/linked_list.lua
local Node = {}
Node.__index = Node

function Node.new(value)
  local self = setmetatable({}, Node)
  self.prev = nil
  self.value = value
  self.next = nil
  return self
end

local LinkedList = {}
LinkedList.__index = LinkedList
LinkedList.__tostring = function(self)
  local result = "LinkedList: {\n"
  local current = self.head
  while current do
    if not current.value.id then
      result = result .. "\27[1m"
    end
    result = result .. tostring(current.value)
    if not current.value.id then
      result = result .. "\27[0m"
    end
    current = current.next
    if current ~= nil then
      result = result .. ",\n"
    end
  end
  result = result .. "\n}"
  return result
end

function LinkedList.new()
  local self = setmetatable({}, LinkedList)
  self.head = nil
  self.tail = nil
  return self
end

function LinkedList:pop()
  if self.tail == nil then
    return nil
  end

  local value = self.tail.value
  self.tail = self.tail.prev
  self.tail.next = nil
  return value
end

function LinkedList:append(value)
  local node = Node.new(value)
  if not self.head then
    self.head = node
    self.tail = node
  else
    node.prev = self.tail
    self.tail.next = node
    self.tail = node
  end
end

function LinkedList:insert(value, after)
  if after.next == nil then
    return self:append(value)
  end
  local node = Node.new(value)

  local tmp = after.next
  tmp.prev = node

  node.next = tmp
  node.prev = after
  after.next = node
end

function LinkedList:remove(node)
  if node.next == nil then
    self.tail = self.tail.prev
    self.tail.next = nil
  else
    local next = node.next
    local prev = node.prev

    prev.next = next
    if next ~= nil then
      next.prev = prev
    end
  end
end

return LinkedList

File: lib/collection/linked-list/linked_list.ts
import { Node } from "./node.js";

export class LinkedList<Element> {
  head: Node<Element> | null;
  // @todo should this be a WeakRef?
  tail: Node<Element> | null;

  constructor() {
    this.head = null;
    this.tail = null;
  }

  /**
    * Create a iterator from the list.
    *
    * Allows the `LinkedList` to be used in for loops etc.
    */
  *[Symbol.iterator](): Iterator<Node<Element>> {
    let current = this.head;
    while (current != null) {
      yield current;

      current = current.next;
    }
  }

  /**
    * Returns if the list is empty
    */
  get isEmpty(): boolean {
    return this.head == null
  }

  /**
    * Append the `element` to the end of the list.
    *
    * @time: O(1)
    * @memory: O(1)
    */
  append(element: Element) {
    // Note: not using `isEmpty` so Typescript can statically see that tail is
    // defined.
    if (!this.tail) {
      this.push(element)
    } else {
      // Make a new tail and set the previous to the current tail.
      const node = new Node(this.tail, element, null)

      this.tail.next = node;
      this.tail = node;
    }
  }

  /**
    * Push the `element` to the head of the list.
    *
    * @time: O(1)
    * @memory: O(1)
    */
  push(element: Element) {
    const node = new Node(null, element, null)

    if (this.head) {
      this.head.previous = node;
      node.next = this.head;
      this.head = node;
    } else {
      this.head = node
      this.tail = node
    }
  }

  /**
    * Pop the tail of the list and return it.
    *
    * @time: O(1)
    * @memory: O(1)
    */
  pop(): Node<Element> | null {
    const node = this.tail;

    // The new tail will be the currents tail previous. We basically move the
    // pointer back.
    this.tail = this.tail?.previous ?? null;
    if (this.tail) {
      this.tail.next = null;
    } else {
      this.head = null;
    }

    return node;
  }
}
